faunadb/query.py

```diff
@@ -85,6 +88,48 @@ def _to_lambda(value):
     return value
 
 
+def lambda_pattern(pattern, lambda_body):
+  """
+  See the `docs <https://faunadb.com/documentation/queries#basic_forms>`__.
+  This form gathers variables from the pattern you provide and puts them in an object.
+  It is called like::
+
+      q = query.map_expr(
+        query.lambda_pattern(["foo", "", "bar"], lambda args: [args.bar, args.foo]),
+        [[1, 2, 3], [4, 5, 6]]))
+      # Result of client.query(q) is: [[3, 1], [6, 4]].
+
+  You can also destructure the namedtuple; in this case, variables are sorted alphabetically::
+
+      query.lambda_pattern(["foo", "", "bar"], lambda (bar, foo): [bar, foo])
+
+  :param Sequence|Mapping pattern:
+    Tree of lists and dicts. Leaves are the names of variables.
+    If a leaf is the empty string ``""``, it is ignored.
+  :param function lambda_body:
+    Takes a ``namedtuple`` of variables taken from the leaves of ``pattern``, and returns a query.
+  """
+  var_names = []
+  def collect_vars(pat):
+    if isinstance(pat, str):
+      if pat != "":
+        var_names.append(pat)
+    elif isinstance(pat, Sequence):
+      for entry in pat:
+        collect_vars(entry)
+    elif isinstance(pat, Mapping):
+      for val in pat.values():
+        collect_vars(val)
+    else:
+      raise InvalidQuery, "Pattern must be a str, Sequence, or Mapping; got %s" % repr(pat)
+
+  collect_vars(pattern)
+  var_names.sort()
+  VarsType = namedtuple("vars", var_names)
```

Comment:
Maybe we could call this `PatternVars` or something? Including changing the type name we give to `namedtuple`.